File: src/athena/core/agent/subagent.py
"""多 Agent 管理 — AgentControl 作为 ThreadRuntime/mailbox 的轻量门面。"""

import asyncio
from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from pydantic_ai.messages import ModelRequest, UserPromptPart

from athena.core.agent.agent import AgentContext, AgentOutcome
from athena.core.schemas import AthenaThread, AthenaTurn
from athena.memory.context_manager import ContextManager

if TYPE_CHECKING:
    from athena.core.agent.agent import Agent


@dataclass(slots=True)
class AgentResult:
    agent_id: str
    result_ref: str = ""
    status: str = "completed"


@dataclass(slots=True, frozen=True)
class AgentEvent:
    kind: str
    event_ref: str
    data: dict[str, Any] | None = None


@dataclass(slots=True)
class AgentHandle:
    """子 Agent 句柄 — future + 事件队列 + 内存引用。"""

    agent_id: str
    _future: asyncio.Future[AgentResult] = field(default_factory=asyncio.Future)
    _task: asyncio.Task[None] | None = None
    _memory: "ContextManager | None" = None
    _events: asyncio.Queue[AgentEvent] = field(default_factory=asyncio.Queue)

    async def wait(self, *, timeout: float | None = None) -> AgentResult:
        """等待子 Agent 完成。shield 防止父任务取消传播到子 Agent future。"""
        return await asyncio.wait_for(asyncio.shield(self._future), timeout=timeout)

    async def next_event(self, *, timeout: float | None = None) -> AgentEvent:
        return await asyncio.wait_for(self._events.get(), timeout=timeout)

    def cancel(self) -> None:
        if self._task and not self._task.done():
            self._task.cancel()


class AgentControl:
    """子 Agent 管理器 — Semaphore 限制并发，独立 ContextManager 隔离内存。"""

    def __init__(self, *, max_concurrency: int = 8) -> None:
        self._semaphore = asyncio.Semaphore(max_concurrency)
        self._handles: dict[str, AgentHandle] = {}
        self._counter = 0

    async def spawn(self, agent: "Agent", task: str, *, emit=None) -> AgentHandle:
        """创建子 Agent 并启动后台执行。Semaphore 控制并发数。"""
        await self._semaphore.acquire()
        self._counter += 1
        agent_id = f"sub-{self._counter}"
        handle = AgentHandle(agent_id=agent_id, _memory=ContextManager())
        self._handles[agent_id] = handle
        handle._task = asyncio.create_task(
            self._run(agent_id, agent, task, handle, emit)
        )
        return handle

    async def _run(
        self, agent_id: str, agent: "Agent", task: str, handle: AgentHandle, emit
    ) -> None:
        """子 Agent 执行体：构造合成 Thread/Turn → agent.run() → 解析 future。

        Semaphore 在 spawn 时获取，在 finally 中释放，保证最大并发数。
        """
        try:
            memory = handle._memory
            assert memory is not None

            async def _emit(
                kind: str, event_ref: str, data: dict[str, Any] | None = None
            ) -> None:
                # 双通道：写入 handle 事件队列 + 转发到父级 emit
                await handle._events.put(AgentEvent(kind, event_ref, data))
                if emit is not None:
                    await emit(kind, event_ref, data)

            thread = AthenaThread(
                thread_id=agent_id,
                session_id="sub",
                status="running",
                context_ref=f"sub://{agent_id}",
            )
            turn = AthenaTurn(
                turn_id=f"{agent_id}-1",
                thread_id=agent_id,
                request_ref=task or "task",
                status="running",
            )
            ctx = AgentContext(
                thread, turn, _emit, agent.config.tools, asyncio.Event(), memory
            )

            outcome: AgentOutcome = await agent.run(ctx)
            handle._future.set_result(
                AgentResult(agent_id=agent_id, result_ref=outcome.result_ref)
            )
        except asyncio.CancelledError:
            # AgentHandle.cancel() 或外部中断 → 标记为 cancelled 而非异常
            if not handle._future.done():
                handle._future.set_result(
                    AgentResult(agent_id=agent_id, status="cancelled")
                )
        except Exception as exc:
            if not handle._future.done():
                handle._future.set_exception(exc)
        finally:
            self._semaphore.release()
            self._handles.pop(agent_id, None)

    async def send_message(self, agent_id: str, message: str) -> None:
        """向运行中的子 Agent 发送消息，追加到其 ContextManager。"""
        handle = self._handles.get(agent_id)
        if handle is None or handle._memory is None:
            raise KeyError(f"unknown agent: {agent_id}")
        handle._memory.append(ModelRequest(parts=[UserPromptPart(content=message)]))

    async def interrupt(self, agent_id: str) -> None:
        h = self._handles.get(agent_id)
        if h:
            h.cancel()

    def list_agents(self) -> list[str]:
        return list(self._handles.keys())

File: athena-rust/crates/athena-agent/src/subagent.rs
use crate::agent::{Agent, AgentContext, AgentError};
use athena_memory::{ContextManager, MessagePart, MessageRole, ModelMessage};
use athena_runtime::{EmitError, EventDraft, EventSink};
use athena_types::{
    AthenaThread, AthenaTurn, SessionId, ThreadId, ThreadStatus, TurnId, TurnStatus,
};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use std::sync::atomic::{AtomicUsize, Ordering};
use tokio::sync::{Mutex, Semaphore, mpsc, watch};
use tokio_util::sync::CancellationToken;

const SUB_CONTEXT_LIMIT: usize = 128_000;

/// A sub-agent's structured result.
#[derive(Debug, Clone)]
pub struct AgentResult {
    pub agent_id: String,
    pub result_ref: String,
    pub status: String,
}

/// An event forwarded from a running sub-agent.
#[derive(Debug, Clone)]
pub struct AgentEvent {
    pub kind: String,
    pub event_ref: String,
    pub data: Option<Value>,
}

/// Handle to a spawned sub-agent: its result, event stream, and cancellation.
pub struct AgentHandle {
    pub agent_id: String,
    result: watch::Receiver<Option<AgentResult>>,
    events: Mutex<mpsc::UnboundedReceiver<AgentEvent>>,
    cancel: CancellationToken,
}

impl AgentHandle {
    /// Await the sub-agent's terminal result.
    pub async fn wait(&self) -> AgentResult {
        let mut rx = self.result.clone();
        loop {
            if let Some(result) = rx.borrow().clone() {
                return result;
            }
            if rx.changed().await.is_err() {
                return AgentResult {
                    agent_id: self.agent_id.clone(),
                    result_ref: String::new(),
                    status: "unknown".into(),
                };
            }
        }
    }

    /// Receive the next event, or `None` once the sub-agent finished.
    pub async fn next_event(&self) -> Option<AgentEvent> {
        self.events.lock().await.recv().await
    }

    pub fn cancel(&self) {
        self.cancel.cancel();
    }
}

struct Entry {
    memory: Arc<Mutex<ContextManager>>,
    cancel: CancellationToken,
}

/// Manages sub-agents with a concurrency limit and isolated memory per agent.
pub struct AgentControl {
    sem: Arc<Semaphore>,
    inner: Arc<Mutex<HashMap<String, Entry>>>,
    counter: AtomicUsize,
}

impl AgentControl {
    pub fn new(max_concurrency: usize) -> Self {
        Self {
            sem: Arc::new(Semaphore::new(max_concurrency.max(1))),
            inner: Arc::new(Mutex::new(HashMap::new())),
            counter: AtomicUsize::new(0),
        }
    }

    /// Spawn a sub-agent running `task`, optionally forwarding events to `parent`.
    pub async fn spawn(
        &self,
        agent: Arc<Agent>,
        task: String,
        parent: Option<Arc<dyn EventSink>>,
    ) -> Result<AgentHandle, AgentError> {
        let permit = self
            .sem
            .clone()
            .acquire_owned()
            .await
            .map_err(|_| AgentError::Invalid("agent control closed".into()))?;

        let id = self.counter.fetch_add(1, Ordering::SeqCst) + 1;
        let agent_id = format!("sub-{id}");
        let memory = Arc::new(Mutex::new(ContextManager::new(SUB_CONTEXT_LIMIT)));
        let cancel = CancellationToken::new();

        let thread = AthenaThread {
            thread_id: id_err(ThreadId::new(agent_id.clone()))?,
            session_id: id_err(SessionId::new("sub"))?,
            status: ThreadStatus::Running,
            context_ref: id_err(athena_types::ArtifactRef::new(format!("sub://{agent_id}")))?,
        };
        let request = if task.is_empty() {
            "task".to_string()
        } else {
            task
        };
        let turn = AthenaTurn {
            turn_id: id_err(TurnId::new(format!("{agent_id}-1")))?,
            thread_id: thread.thread_id.clone(),
            request_ref: id_err(athena_types::ArtifactRef::new(request))?,
            status: TurnStatus::Running,
            result_ref: None,
        };

        self.inner.lock().await.insert(
            agent_id.clone(),
            Entry {
                memory: memory.clone(),
                cancel: cancel.clone(),
            },
        );

        let (result_tx, result_rx) = watch::channel(None);
        let (events_tx, events_rx) = mpsc::unbounded_channel();
        let sink: Arc<dyn EventSink> = Arc::new(SubAgentSink {
            events: events_tx,
            parent,
        });
        let inner = self.inner.clone();
        let run_cancel = cancel.clone();
        let run_id = agent_id.clone();

        tokio::spawn(async move {
            let (_ctx_cancel_tx, ctx_cancel_rx) = watch::channel(false);
            let ctx = AgentContext {
                thread,
                turn,
                emit: sink,
                cancel: ctx_cancel_rx,
                memory,
            };
            let result = tokio::select! {
                r = agent.run(ctx) => match r {
                    Ok(o) => AgentResult { agent_id: run_id.clone(), result_ref: o.result_ref.as_str().to_string(), status: "completed".into() },
                    Err(_) => AgentResult { agent_id: run_id.clone(), result_ref: String::new(), status: "failed".into() },
                },
                _ = run_cancel.cancelled() => AgentResult { agent_id: run_id.clone(), result_ref: String::new(), status: "cancelled".into() },
            };
            let _ = result_tx.send(Some(result));
            drop(permit);
            inner.lock().await.remove(&run_id);
        });

        Ok(AgentHandle {
            agent_id,
            result: result_rx,
            events: Mutex::new(events_rx),
            cancel,
        })
    }

    /// Append a message to a running sub-agent's context.
    pub async fn send_message(&self, agent_id: &str, message: &str) -> Result<(), AgentError> {
        let memory = self
            .inner
            .lock()
            .await
            .get(agent_id)
            .map(|e| e.memory.clone());
        match memory {
            Some(mem) => {
                mem.lock().await.append(ModelMessage {
                    role: MessageRole::User,
                    parts: vec![MessagePart::UserPrompt {
                        content: message.to_string(),
                    }],
                });
                Ok(())
            }
            None => Err(AgentError::Invalid(format!("unknown agent: {agent_id}"))),
        }
    }

    pub async fn interrupt(&self, agent_id: &str) {
        if let Some(entry) = self.inner.lock().await.get(agent_id) {
            entry.cancel.cancel();
        }
    }

    pub async fn list_agents(&self) -> Vec<String> {
        self.inner.lock().await.keys().cloned().collect()
    }
}

fn id_err<T>(r: Result<T, athena_types::ValidationError>) -> Result<T, AgentError> {
    r.map_err(|e| AgentError::Invalid(e.to_string()))
}

struct SubAgentSink {
    events: mpsc::UnboundedSender<AgentEvent>,
    parent: Option<Arc<dyn EventSink>>,
}

#[async_trait::async_trait]
impl EventSink for SubAgentSink {
    async fn emit(&self, draft: EventDraft) -> Result<(), EmitError> {
        let _ = self.events.send(AgentEvent {
            kind: draft.kind.clone(),
            event_ref: draft.event_ref.clone(),
            data: draft.data.clone(),
        });
        if let Some(parent) = &self.parent {
            let _ = parent.emit(draft).await;
        }
        Ok(())
    }
}
